File: prefig/core/legend.py
import lxml.etree as ET
import copy
import numpy as np
from . import user_namespace as un
from . import utilities as util
from . import tags
from . import label
from . import point
from . import CTM


# This feels a little hack-y and I'm somewhat ambivalent about legends
# as they may not be a good practice for tactile diagrams.

class Legend:
    def __init__(self, element, diagram, parent, outline_status):
        self.element = element
        self.diagram = diagram
        self.parent = parent
        self.outline = outline_status == 'add_outline'
        self.tactile = diagram.output_format() == 'tactile'

        # we will put everything in a group and apply a final transform to it
        self.group = ET.Element('g')
        parent.append(self.group)

        # register this legend so we an come back to it
        diagram.add_legend(self)

        # let's get some basic data about the legend
        # self.def_anchor holds the SVG coordinates of the anchor
        anchor_str = element.get('anchor', '(bbox[2],bbox[3])')
        user_anchor = un.valid_eval(anchor_str)
        self.def_anchor = diagram.transform(user_anchor)
        
        alignment = element.get('alignment')
        if alignment == 'e':
            element.set('alignment', 'east')
        alignment = util.get_attr(element, 'alignment', 'c')
        self.displacement = label.alignment_displacement[alignment]

        # let's go through each of the child and set up the labels
        self.li_dict = {}
        key_width = 0
        point_width = 10

        # length of line as a key
        self.line_width = 24
        if self.tactile:
            self.line_width = 45

        dummy_group = ET.Element('g')
        for num, li in enumerate(element):
            if li.tag != 'item':
                print(f"{li.tag} is not allowed inside a <legend>")
                continue

            # first we'll create the label
            label_el = copy.deepcopy(li)
            label_el.tag = 'label'
            label_el.set('id', f"legend-label-{num}")
            label_el.set('alignment', 'se')
            label_el.set('anchor', element.get('anchor', anchor_str))
            label_el.set('abs-offset', '(0,0)')
            label_el.set('justify', 'left')
            label.label(label_el, diagram, dummy_group, None)
            dummy_group.clear()
            
            # and now the key
            ref = label_el.get('ref')
            search = f'//*[@id="{ref}"]'
            references = diagram.diagram_element.xpath(search)
            if len(references) != 1:
                print(f"{ref} should refer to exactly one element")
                continue
            key = references[0]
            if key.tag == 'point':
                key = copy.deepcopy(key)
                key.set('p', anchor_str)
                key.set('size', '4')
                key.set('id', f"legend-point-{num}")
                key_width = max(key_width, point_width)
            else:
                fill =  key.get('fill')
                if fill is None or fill == 'none':
                    key_el = ET.Element('line')
                    key_el.set('stroke', key.get('stroke'))
                    dash = key.get('dash', None)
                    if dash is not None:
                        key_el.set('stroke-dasharray', dash)
                    key_width = max(key_width, self.line_width)
                    key = key_el
                else:
                    key_el = ET.Element('point')
                    key_el.set('stroke', key.get('stroke'))
                    key_el.set('fill', fill)
                    key_el.set('style', key.get('style', 'box'))
                    key_el.set('size', '5')
                    key_width = max(key_width, point_width)
                    key = key_el

            self.li_dict[li] = [key, label_el]
        self.key_width = key_width

    def place_legend(self, diagram):
        # We need to revisit the legend after the labels are created
        # and set their positions

        # Labels in tactile diagrams have their own properties
        if self.tactile:
            self.place_tactile_legend(diagram)
            return

        # We're doing this at the very end so the diagram.ctm is the default
        outer_padding = 5
        center_padding = 10
        interline = un.valid_eval(self.element.get('vertical-skip', '7'))
        height = outer_padding
        label_width = 0

        # Determine the width of the labels and the height of the entire legend
        for li in self.element:
            key, label = self.li_dict[li]
            label_dims = self.diagram.get_label_dims(label)
            height += label_dims[1] + interline
            label_width = max(label_width, label_dims[0])
        height += outer_padding - interline
            
        width = label_width + 2*outer_padding + self.key_width + center_padding

        # offset from the anchor
        offset = [8*(self.displacement[0] + 0.5),
                  8*(self.displacement[1]-0.5)]

        # set up a translate for the legend's bounding box
        p = self.def_anchor
        tform = CTM.translatestr(p[0] + offset[0], p[1] - offset[1])
        scale = un.valid_eval(self.element.get('scale', '1'))
        dx = scale * width * self.displacement[0]
        dy = scale * height * self.displacement[1]
        tform = tform + ' ' + CTM.translatestr(dx, -dy)

        tform = tform + ' ' + CTM.scalestr(scale, scale)
        self.group.set('transform', tform)

        # The following component is the bounding box of the legend
        rect = ET.Element('rect')
        self.group.append(rect)
        rect.set('x', '0')
        rect.set('y', '0')
        rect.set('width', str(width))
        rect.set('height', str(height))
        rect.set('stroke', self.element.get('stroke', 'black'))
        rect.set('fill', 'white')
        rect.set('fill-opacity', self.element.get('opacity', '1'))

        # Now we'll go through and place the labels and keys
        label_x = outer_padding + self.key_width + center_padding
        y = outer_padding
        for li in self.element:
            key, label = self.li_dict[li]
            label_dims = self.diagram.get_label_dims(label)
            label_group = self.diagram.get_label_group(label)[0]
            tform = CTM.translatestr(label_x, y)
            label_group.set('transform', tform)
            self.group.append(label_group)

            key_y = y + label_dims[1]/2

            if key.tag == 'point':
                key_x = outer_padding + self.key_width/2
                user_point = self.diagram.inverse_transform((key_x,key_y))
                key.set('p', util.pt2str(user_point, spacer=","))
                point.point(key, self.diagram, self.group, None)
            if key.tag == 'line':
                key_x0 = outer_padding
                key_x1 = outer_padding + self.line_width
                key.set('x1', str(key_x1))
                key.set('y1', str(key_y))
                key.set('x2', str(key_x0))
                key.set('y2', str(key_y))
                key.set('stroke-width', '2')
                self.group.append(key)

            y += label_dims[1] + interline

        
    def place_tactile_legend(self, diagram):
        # This is the same outline as above but works for tactile labels
        
        # The labels in a tactile diagram have already been placed
        # in the SVG so we first need to remove them
        root = self.diagram.get_root()
        groups = root.findall('g')
        label_groups = []
        for group in groups:
            id = group.get('id', 'none')
            if id == 'background-group':
                for rectangle in group:
                    if rectangle.get('id', 'none').startswith('legend-label'):
                        group.remove(rectangle)
            if id == 'braille-group':
                for label in group:
                    if label.get('id','none').startswith('legend-label'):
                        label_groups.append(label)
                        group.remove(label)

        # We're doing this at the very end so the diagram.ctm is the default
        
        gap = 3.6 # gap between embossed dots
        outer_padding = 3*gap # 1/8th of an inch
        center_padding = 6*gap

        # TODO:  This feature needs to be checked with users
        # It's not clear what this spacing should be
        interline = 4 * gap  # the vertical space between two items
        height = outer_padding
        label_width = 0

        # Find dimensions of the bounding box
        for li in self.element:
            key, label = self.li_dict[li]
            label_dims = self.diagram.get_label_dims(label)
            height += label_dims[1] + interline
            label_width = max(label_width, label_dims[0])
        height += outer_padding - interline
            
        width = label_width + 2*outer_padding + self.key_width + center_padding
        
        # Get the offset
        offset = [8*(self.displacement[0] + 0.5),
                  8*(self.displacement[1]-0.5)]
        
        offset = [o + 6*np.sign(o) for o in offset]
        if self.displacement[0] == 0:
            offset[0] += 6
        if self.displacement[1] == -1:
            offset[1] -= 6

        # Set up the translation for the bounding box    
        p = self.def_anchor
        dx = width * self.displacement[0]
        dy = height * self.displacement[1]

        translate = (p[0] + offset[0] + dx,
                     p[1] - offset[1] - dy)
        translate = [gap * round(c/gap) for c in translate]
        tform = CTM.translatestr(*translate)

        self.group.set('transform', tform)

        # Here is the bounding box
        rect = ET.Element('rect')
        self.group.append(rect)
        rect.set('x', '0')
        rect.set('y', '0')
        rect.set('width', str(width))
        rect.set('height', str(height))
        rect.set('stroke', self.element.get('stroke', 'black'))
        rect.set('fill', 'white')
            
        # Now place the labels and keys
        label_x = outer_padding + self.key_width + center_padding
        y = outer_padding
        for num, li in enumerate(self.element):
            key, label = self.li_dict[li]
            label_dims = self.diagram.get_label_dims(label)
            label_group = self.diagram.get_label_group(label)[0]
            label_x = gap * round(label_x/gap)
            y = gap * round(y/gap)
            tform = CTM.translatestr(label_x, y)
            label_group.set('transform', tform)
            label = label_groups[num]
            label_height = gap * round(label_dims[1]/gap)
            tform = CTM.translatestr(0, label_height)
            label.set('transform', tform)
            label_group.append(label_groups[num])
            self.group.append(label_group)

            key_y = y + label_dims[1]/2

            if key.tag == 'point':
                key_x = outer_padding + self.key_width/2
                user_point = self.diagram.inverse_transform((key_x,key_y))
                key.set('p', util.pt2str(user_point, spacer=","))
                point.point(key, self.diagram, self.group, None)
            if key.tag == 'line':
                key_x0 = outer_padding
                key_x1 = outer_padding + self.line_width
                key.set('x1', str(key_x1))
                key.set('y1', str(key_y))
                key.set('x2', str(key_x0))
                key.set('y2', str(key_y))
                key.set('stroke-width', '2')
                self.group.append(key)

            y += label_dims[1] + interline

        
def legend(element, diagram, parent, outline_status):
    
    # if we're finishing outline, we'll wait until
    # we process the legend at the end
    if outline_status == 'finish_outline':
        return

    Legend(element, diagram, parent, outline_status)

File: packages/prefig-rust/prefig-core/src/core/legend.rs
//! Port of prefig/core/legend.py (non-tactile path; tactile legends TODO).

use crate::core::ctm;
use crate::core::diagram::Diagram;
use crate::core::utilities::{self as util, pt2str};
use crate::core::{label, point};
use crate::value::py_str;
use crate::xml::{self, El};

pub struct LegendData {
    pub element: El,
    pub group: El,
    pub def_anchor: [f64; 2],
    pub displacement: [f64; 2],
    /// (item element, key element, label element)
    pub items: Vec<(El, El, El)>,
    pub key_width: f64,
    pub line_width: f64,
}

pub fn legend(element: &El, diagram: &mut Diagram, parent: &El, _outline_group: Option<&El>) {
    let tactile = diagram.output_format() == "tactile";

    let group = xml::new_element("g");
    let id = element.borrow().get("id");
    diagram.add_id(&group, id.as_deref());
    xml::append(parent, &group);
    diagram.register_svg_element(element, &group);

    let anchor_str = element.borrow().get_or("anchor", "(bbox[2],bbox[3])");
    let Some(user_anchor) = diagram
        .ctx
        .valid_eval(&anchor_str)
        .ok()
        .and_then(|v| v.as_vec_f64().ok())
    else {
        log::error!("Error in <legend> evaluating anchor={anchor_str}");
        return;
    };
    let def_anchor = diagram.transform([user_anchor[0], user_anchor[1]]);

    if element.borrow().get_or("alignment", "c") == "e" {
        element.borrow_mut().set("alignment", "east");
    }
    let alignment = util::get_attr(element, "alignment", "c", &mut diagram.ctx);
    let displacement = label::alignment_displacement(&alignment).unwrap_or([-0.5, 0.5]);

    let mut items: Vec<(El, El, El)> = Vec::new();
    let mut key_width = 0.0f64;
    let point_width = 10.0;
    let line_width = if tactile { 72.0 } else { 24.0 };

    let dummy_group = xml::new_element("g");
    let children: Vec<El> = element.borrow().children.clone();
    for (num, li) in children.iter().enumerate() {
        if li.borrow().tag != "item" {
            log::warn!("{} is not allowed inside a <legend>", li.borrow().tag);
            continue;
        }

        // the label
        let label_id_stub = diagram.prepend_id_prefix("legend-label");
        let label_el = xml::deep_copy(li);
        {
            let mut l = label_el.borrow_mut();
            l.tag = "label".to_string();
            l.set("id", &format!("{label_id_stub}-{num}"));
            l.set("alignment", "se");
            l.set("anchor", &anchor_str);
            l.set("abs-offset", "(0,0)");
            l.set("justify", "left");
        }
        label::label(&label_el, diagram, &dummy_group, None);

        // and the key: find the referenced element by id or at
        let ref_attr = label_el.borrow().get_or("ref", "");
        let mut reference = None;
        for candidate in xml::iter_subtree(&diagram.diagram_element) {
            let c = candidate.borrow();
            if c.get("id").as_deref() == Some(&ref_attr)
                || c.get("at").as_deref() == Some(&ref_attr)
            {
                drop(c);
                reference = Some(candidate);
                break;
            }
        }
        let Some(key_src) = reference else {
            log::warn!("{ref_attr} should refer to an element");
            continue;
        };

        let point_id_stub = diagram.prepend_id_prefix("legend-point");
        let key = if key_src.borrow().tag == "point" {
            let key = xml::deep_copy(&key_src);
            {
                let mut k = key.borrow_mut();
                k.set("p", &anchor_str);
                k.set("size", "4");
                k.set("id", &format!("{point_id_stub}-{num}"));
            }
            key_width = key_width.max(point_width);
            key
        } else {
            let fill = key_src.borrow().get("fill");
            if fill.is_none() || fill.as_deref() == Some("none") {
                let key_el = xml::new_element("line");
                {
                    let mut k = key_el.borrow_mut();
                    k.set("stroke", &key_src.borrow().get_or("stroke", "none"));
                    if let Some(dash) = key_src.borrow().get("dash") {
                        k.set("stroke-dasharray", &dash);
                    }
                }
                key_width = key_width.max(line_width);
                key_el
            } else {
                let key_el = xml::new_element("point");
                {
                    let mut k = key_el.borrow_mut();
                    k.set("stroke", &key_src.borrow().get_or("stroke", "none"));
                    k.set("fill", &fill.unwrap_or_default());
                    k.set("style", &key_src.borrow().get_or("style", "box"));
                    k.set("size", "5");
                }
                key_width = key_width.max(point_width);
                key_el
            }
        };

        items.push((li.clone(), key, label_el));
    }

    diagram.legends.push(LegendData {
        element: element.clone(),
        group,
        def_anchor,
        displacement,
        items,
        key_width,
        line_width,
    });
}

pub fn place_legend(diagram: &mut Diagram, data: &LegendData) {
    if diagram.output_format() == "tactile" {
        place_tactile_legend(diagram, data);
        return;
    }

    let outer_padding = 5.0;
    let center_padding = 10.0;
    let interline_attr = data.element.borrow().get_or("vertical-skip", "7");
    let interline = diagram
        .ctx
        .valid_eval(&interline_attr)
        .ok()
        .and_then(|v| v.as_num().ok())
        .unwrap_or(7.0);
    let mut height = outer_padding;
    let mut label_width = 0.0f64;

    for (_, _, label_el) in &data.items {
        let Some(dims) = diagram.get_label_dims(label_el) else {
            log::warn!("There is a missing label in a <legend>");
            continue;
        };
        height += dims.1 + interline;
        label_width = label_width.max(dims.0);
    }
    height += outer_padding - interline;

    let width = label_width + 2.0 * outer_padding + data.key_width + center_padding;

    let offset = [
        8.0 * (data.displacement[0] + 0.5),
        8.0 * (data.displacement[1] - 0.5),
    ];
    let p = data.def_anchor;
    let mut tform = ctm::translatestr(p[0] + offset[0], p[1] - offset[1]);
    let scale_attr = data.element.borrow().get_or("scale", "1");
    let scale = diagram
        .ctx
        .valid_eval(&scale_attr)
        .ok()
        .and_then(|v| v.as_num().ok())
        .unwrap_or(1.0);
    let dx = scale * width * data.displacement[0];
    let dy = scale * height * data.displacement[1];
    tform = format!("{tform} {}", ctm::translatestr(dx, -dy));
    tform = format!("{tform} {}", ctm::scalestr(scale, scale));
    data.group.borrow_mut().set("transform", &tform);

    // the legend's bounding box
    let rect = xml::new_element("rect");
    xml::append(&data.group, &rect);
    {
        let mut r = rect.borrow_mut();
        r.set("x", "0");
        r.set("y", "0");
        r.set("width", &py_str(width));
        r.set("height", &py_str(height));
        r.set("stroke", &data.element.borrow().get_or("stroke", "black"));
        r.set("fill", "white");
        r.set(
            "fill-opacity",
            &data.element.borrow().get_or("opacity", "1"),
        );
    }

    // place the labels and keys
    let label_x = outer_padding + data.key_width + center_padding;
    let mut y = outer_padding;
    for (_, key, label_el) in &data.items {
        let Some(dims) = diagram.get_label_dims(label_el) else {
            continue;
        };
        let Some(label_group) = diagram.get_label_group(label_el) else {
            continue;
        };
        let tform = ctm::translatestr(label_x, y);
        label_group.borrow_mut().set("transform", &tform);
        xml::append(&data.group, &label_group);

        let key_y = y + dims.1 / 2.0;
        let key_tag = key.borrow().tag.clone();
        if key_tag == "point" {
            let key_x = outer_padding + data.key_width / 2.0;
            let user_point = diagram.inverse_transform([key_x, key_y]);
            key.borrow_mut().set("p", &pt2str(user_point, ","));
            point::point(key, diagram, &data.group, None);
        }
        if key_tag == "line" {
            let key_x0 = outer_padding;
            let key_x1 = outer_padding + data.line_width;
            {
                let mut k = key.borrow_mut();
                k.set("x1", &py_str(key_x1));
                k.set("y1", &py_str(key_y));
                k.set("x2", &py_str(key_x0));
                k.set("y2", &py_str(key_y));
                k.set("stroke-width", "2");
            }
            xml::append(&data.group, key);
        }

        y += dims.1 + interline;
    }
}

/// Port of legend.py place_tactile_legend: braille labels were already emitted
/// into the diagram; move them into the legend group and lay them out on the
/// embossing grid.
fn place_tactile_legend(diagram: &mut Diagram, data: &LegendData) {
    let gap = 3.6;

    // the braille labels were placed into background-group / braille-group;
    // pull the ones belonging to this legend back out
    let legend_label_id = diagram.prepend_id_prefix("legend-label");
    let mut label_groups: Vec<El> = Vec::new();
    let root_children: Vec<El> = diagram.root.borrow().children.clone();
    for g in &root_children {
        let id = g.borrow().get_or("id", "none");
        if id == "background-group" {
            let rects: Vec<El> = g.borrow().children.clone();
            for rect in rects {
                if rect
                    .borrow()
                    .get_or("id", "none")
                    .starts_with(&legend_label_id)
                {
                    xml::remove(g, &rect);
                }
            }
        }
        if id == "braille-group" {
            let labels: Vec<El> = g.borrow().children.clone();
            for label in labels {
                if label
                    .borrow()
                    .get_or("id", "none")
                    .starts_with(&legend_label_id)
                {
                    label_groups.push(label.clone());
                    xml::remove(g, &label);
                }
            }
        }
    }

    let outer_padding = 4.2 * gap;
    let center_padding = 6.0 * gap;
    let interline = 4.0 * gap;
    let mut height = outer_padding;
    let mut label_width = 0.0f64;
    for (_, _, label_el) in &data.items {
        let Some(dims) = diagram.get_label_dims(label_el) else {
            continue;
        };
        height += dims.1 + interline;
        label_width = label_width.max(dims.0);
    }
    height += outer_padding - interline;
    let width = label_width + 2.0 * outer_padding + data.key_width + center_padding;

    let mut offset = [
        8.0 * (data.displacement[0] + 0.5),
        8.0 * (data.displacement[1] - 0.5),
    ];
    offset = [
        offset[0] + 6.0 * offset[0].signum(),
        offset[1] + 6.0 * offset[1].signum(),
    ];
    if data.displacement[0] == 0.0 {
        offset[0] += 6.0;
    }
    if data.displacement[1] == -1.0 {
        offset[1] -= 6.0;
    }

    let p = data.def_anchor;
    let dx = width * data.displacement[0];
    let dy = height * data.displacement[1];
    let translate = [
        gap * ((p[0] + offset[0] + dx) / gap).round_ties_even(),
        gap * ((p[1] - offset[1] - dy) / gap).round_ties_even(),
    ];
    data.group
        .borrow_mut()
        .set("transform", &ctm::translatestr(translate[0], translate[1]));

    let rect = xml::sub_element(&data.group, "rect");
    {
        let mut r = rect.borrow_mut();
        r.set("x", "0");
        r.set("y", "0");
        r.set("width", &py_str(width));
        r.set("height", &py_str(height));
        r.set("stroke", &data.element.borrow().get_or("stroke", "black"));
        r.set("fill", "white");
    }

    let label_x = outer_padding + data.key_width + center_padding;
    let mut y = outer_padding;
    for (num, (_, key, label_el)) in data.items.iter().enumerate() {
        let Some(dims) = diagram.get_label_dims(label_el) else {
            continue;
        };
        let Some(label_group) = diagram.get_label_group(label_el) else {
            continue;
        };
        let label_x = gap * (label_x / gap).round_ties_even();
        y = gap * (y / gap).round_ties_even();
        label_group
            .borrow_mut()
            .set("transform", &ctm::translatestr(label_x, y));
        if let Some(braille) = label_groups.get(num) {
            let label_height = gap * (dims.1 / gap).round_ties_even();
            braille
                .borrow_mut()
                .set("transform", &ctm::translatestr(0.0, label_height));
            xml::append(&label_group, braille);
        }
        xml::append(&data.group, &label_group);

        let key_y = y + dims.1 / 2.0;
        let key_tag = key.borrow().tag.clone();
        if key_tag == "point" {
            let key_x = outer_padding + data.key_width / 2.0;
            let user_point = diagram.inverse_transform([key_x, key_y]);
            key.borrow_mut().set("p", &pt2str(user_point, ","));
            point::point(key, diagram, &data.group, None);
        }
        if key_tag == "line" {
            let key_x0 = outer_padding;
            let key_x1 = outer_padding + data.line_width;
            {
                let mut k = key.borrow_mut();
                k.set("x1", &py_str(key_x1));
                k.set("y1", &py_str(key_y));
                k.set("x2", &py_str(key_x0));
                k.set("y2", &py_str(key_y));
                k.set("stroke-width", "2");
            }
            xml::append(&data.group, key);
        }
        y += dims.1 + interline;
    }
}
